ledger/smt.go

```diff
@@ -0,0 +1,399 @@
+	s.update(rnode, keys, values, batch, 2*iBatch+2, height-1, shortcut, store, newch)
+	res := <-newch
+	if res.err != nil {
+		ch <- result{nil, res.err}
+		return
+	}
+	ch <- result{s.interiorHash(lnode, res.update, height, iBatch, root, shortcut, store, keys, values,
+		batch), nil}
+}
+
+// updateLeft updates the left side of the tree
+func (s *SMT) updateLeft(lnode, rnode, root []byte, keys, values, batch [][]byte, iBatch, height int, shortcut,
+	store bool, ch chan<- result) {
+	// all the keys go in the left subtree
+	newch := make(chan result, 1)
+	s.update(lnode, keys, values, batch, 2*iBatch+1, height-1, shortcut, store, newch)
+	res := <-newch
+	if res.err != nil {
+		ch <- result{nil, res.err}
+		return
+	}
+	ch <- result{s.interiorHash(res.update, rnode, height, iBatch, root, shortcut, store, keys, values,
+		batch), nil}
+}
+
+// splitKeys devides the array of keys into 2 so they can update left and right branches in parallel
+func (s *SMT) splitKeys(keys [][]byte, height int) ([][]byte, [][]byte) {
+	for i, key := range keys {
+		if bitIsSet(key, height) {
+			return keys[:i], keys[i:]
+		}
+	}
+	return keys, nil
+}
+
+// maybeAddShortcutToKV adds a shortcut key to the keys array to be updated.
+// this is used when a subtree containing a shortcut node is being updated
+func (s *SMT) maybeAddShortcutToKV(keys, values [][]byte, shortcutKey, shortcutVal []byte) ([][]byte, [][]byte) {
+	newKeys := make([][]byte, 0, len(keys)+1)
+	newVals := make([][]byte, 0, len(keys)+1)
+
+	if bytes.Compare(shortcutKey, keys[0]) < 0 {
+		newKeys = append(newKeys, shortcutKey)
+		newKeys = append(newKeys, keys...)
+		newVals = append(newVals, shortcutVal)
+		newVals = append(newVals, values...)
+	} else if bytes.Compare(shortcutKey, keys[len(keys)-1]) > 0 {
+		newKeys = append(newKeys, keys...)
+		newKeys = append(newKeys, shortcutKey)
+		newVals = append(newVals, values...)
+		newVals = append(newVals, shortcutVal)
+	} else {
+		higher := false
+		for i, key := range keys {
+			if bytes.Equal(shortcutKey, key) {
+				// the shortcut keys is being updated
+				return keys, values
+			}
+			if !higher && bytes.Compare(shortcutKey, key) > 0 {
+				higher = true
+				continue
+			}
+			if higher && bytes.Compare(shortcutKey, key) < 0 {
+				// insert shortcut in slices
+				newKeys = append(newKeys, keys[:i]...)
+				newKeys = append(newKeys, shortcutKey)
+				newKeys = append(newKeys, keys[i:]...)
+				newVals = append(newVals, values[:i]...)
+				newVals = append(newVals, shortcutVal)
+				newVals = append(newVals, values[i:]...)
+				break
+			}
+		}
+	}
+	return newKeys, newVals
+}
+
+// loadChildren looks for the children of a node.
+// if the node is not stored in cache, it will be loaded from db.
+func (s *SMT) loadChildren(root []byte, height, iBatch int, batch [][]byte) ([][]byte, int, []byte, []byte, bool,
+	error) {
+	isShortcut := false
+	if height%4 == 0 {
+		if len(root) == 0 {
+			// create a new default batch
+			batch = make([][]byte, 31)
```

Comment:
Let's get rid of magic numbers like this and use named constants instead.

Thanks.